common/lib/xmodule/xmodule/library_content_module.py

```diff
@@ -371,13 +388,186 @@ def get_child_descriptors(self):
+class AdaptiveLibraryContentModule(AdaptiveLibraryContentFields, LibraryContentModule, AdaptiveLearningAPIMixin):
     """
     A specialized version of Randomized Content Block that communicates
     with an external service to determine when to show its children, and which ones.
     """
 
+    @lazy
+    def parent_course(self):
+        """
+        Return parent course.
+        """
+        return self.runtime.modulestore.get_course(self.course_id, depth=1)
+
+    @lazy
+    def parent_unit_id(self):
+        """
+        Return block ID of parent unit.
+        """
+        parent_unit = self.get_parent()
+        return parent_unit.scope_ids.usage_id.block_id
+
+    @lazy
+    def anonymous_user_id(self):
+        """
+        Return anonymous ID for current user.
+
+        Note that we can't use `user_service.get_anonymous_user_id` here
+        because it only produces meaningful results for staff users.
+        """
+        user_service = self.runtime.service(self, 'user')
+        if user_service:
+            current_user = user_service.get_current_user()
+            user_id = current_user.opt_attrs.get('edx-platform.user_id')
+            anonymous_user_id = self._make_anonymous_user_id(user_id)
+        else:
+            anonymous_user_id = None
+        return anonymous_user_id
+
+    @lazy
+    def child_block_ids(self):
+        """
+        Return list of `block_id`s identifying children of this block.
+        """
+        return [child.block_id for child in self.children]
+
+    @classmethod
+    def make_selection(cls, selected, children, max_count, mode):
+        """
+        Dynamically selects block_ids indicating which of the possible children are displayed to the current user.
+
+        Arguments:
+            selected - list of (block_type, block_id) tuples assigned to this student
+            children - children of this block
+            max_count - number of components to display to each student
+            mode - how content is drawn from the library
+
+        Returns:
+            A dict containing the following keys:
+
+            'selected' (set) of (block_type, block_id) tuples assigned to this student
+            'invalid' (set) of dropped (block_type, block_id) tuples that are no longer valid
+            'overlimit' (set) of dropped (block_type, block_id) tuples that were previously selected
+            'added' (set) of newly added (block_type, block_id) tuples
+
+        Note that this implementation ignores the `max_count` argument:
+        The list of `selected` (block_type, block_id) tuples assigned to this student
+        has been determined based on information from external service providing adaptive learning features,
+        so we only remove blocks that are no longer valid here.
+        """
+        selected = set(tuple(k) for k in selected)  # set of (block_type, block_id) tuples assigned to this student
```

Comment:
@pomegranited The `make_selection` method of the parent class also computes `selected` this way. I tried changing it the way you suggested and it broke existing tests (for the parent class). I'm not sure why, and I don't have room to look into this in detail. So I'll leave this as is for now to at least be consistent. (I do agree that it looks like it should be possible to change the code in the way you suggested; was thinking the same thing when I first read the parent class' code.)